src/librustc_mir/build/mod.rs

```diff
@@ -94,13 +94,18 @@ pub fn mir_build<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) -> Mir<'t
 
             let ty = tcx.type_of(tcx.hir.local_def_id(id));
             let mut abi = fn_sig.abi;
-            let implicit_argument = if let ty::TyClosure(..) = ty.sty {
-                // HACK(eddyb) Avoid having RustCall on closures,
-                // as it adds unnecessary (and wrong) auto-tupling.
-                abi = Abi::Rust;
-                Some((closure_self_ty(tcx, id, body_id), None))
-            } else {
-                None
+            let implicit_argument = match ty.sty {
+                ty::TyClosure(..) => {
+                    // HACK(eddyb) Avoid having RustCall on closures,
+                    // as it adds unnecessary (and wrong) auto-tupling.
+                    abi = Abi::Rust;
+                    Some((closure_self_ty(tcx, id, body_id), None))
+                }
+                ty::TyGenerator(..) => {
+                    let gen_ty =  tcx.body_tables(body_id).node_id_to_type(id);
```

Comment:
It it `Self` until the generator transformation runs, then it's changed to `&mut Self` (which is the signature of `resume`).